maildir/maildir.c

```diff
@@ -1609,6 +1611,57 @@ static enum MailboxType maildir_path_probe(const char *path, const struct stat *
   return MUTT_UNKNOWN;
 }
 
+/**
+ * maildir_check_dir - create a new maildir directory tree
+ * @param path Path to create the maildir under
+ * @retval -1 Creation failed, we attempt to cleanup anything we already created
+ * @retval  0 Creation success
+ *
+ */
+static int maildir_create_maildir(const char *path)
+{
+  char tmp[PATH_MAX];
+
+  for (int i = 0; i < mutt_array_size(maildir_subdirs); i++)
+  {
+    snprintf(tmp, sizeof(tmp), "%s/%s", path, maildir_subdirs[i]);
+
+    if (mutt_file_mkdir(tmp, S_IRWXU) != 0)
+    {
+      mutt_perror(tmp);
+
+      /* Cleanup what we've already made */
+      for (int j = i - 1; j >= 0; j--)
+      {
+        snprintf(tmp, sizeof(tmp), "%s/%s", path, maildir_subdirs[j]);
+        /* Should probably check errors again here... */
+        /* alert user if something went wrong */
+        rmdir(tmp);
+      }
+      rmdir(path);
+      return -1;
+    }
+  }
+  return 0;
+}
+
+/**
+ *  maildir_create_mailbox - Create a maildir mailbox - Implements - MxOps::mbox_create()
+ */
+struct Mailbox *maildir_create_mailbox(struct Account *a, const char *path, const char *name)
+{
+  if (maildir_create_maildir(path) != 0)
+    return NULL;
+  struct Mailbox *m = mailbox_new();
+
+  mutt_buffer_strcpy(&m->pathbuf, path);
+  m->realpath = mutt_str_dup(path);
+  m->name = mutt_str_dup(name);
+  m->type = MUTT_MAILDIR;
```

Comment:
I'm not entirely sure on all the setup we need here, should there also be `mx_ac_add(a, m);` (test-library didn't play nice with this one)